invoices/invoiceregistry.go

```diff
@@ -465,19 +469,67 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 			return nil, errNoUpdate
 		}
 
-		// If the invoice is already canceled, there is no further
-		// checking to do.
-		if inv.Terms.State == channeldb.ContractCanceled {
-			debugLog("invoice already canceled")
+		// Only accept payments to open invoices.
+		if inv.Terms.State != channeldb.ContractOpen {
+			debugLog(fmt.Sprintf(
+				"invoice in state %v and no longer open",
+				inv.Terms.State))
 			return nil, errNoUpdate
 		}
 
-		// If an invoice amount is specified, check that enough
-		// is paid. Also check this for duplicate payments if
-		// the invoice is already settled or accepted.
-		if inv.Terms.Value > 0 && amtPaid < inv.Terms.Value {
-			debugLog("amount too low")
-			return nil, errNoUpdate
+		// Start building the accept descriptor.
+		acceptDesc := &channeldb.HtlcAcceptDesc{
+			Amt:          amtPaid,
+			Expiry:       expiry,
+			AcceptHeight: currentHeight,
+		}
+
+		// If an invoice amount is specified and this is not an mpp,
+		// check that the exact amount is paid.
+		var canSettle bool
+		if mpp == nil {
```

Comment:
do we also need to check if `len(inv.Htlcs) == 0`? my reading atm is that this will also settle any partial htlcs if a regular payment comes thru that settles the invoice.

haven't thought about it too much, but seems we should reject regular payments if we already have some partial mpp payments. if we have an mpp invoice feature bit, this path should also fail if `option_mpp` is required on the invoice

Reply:
good point. added check and todo